File: src/tests/file-c/stat.c
/*
 * Copyright(c) The Maintainers of Nanvix.
 * Licensed under the MIT License.
 */

//==================================================================================================
// Imports
//==================================================================================================

#include <assert.h>
#include <fcntl.h>
#include <stdio.h>
#include <sys/stat.h>

//==================================================================================================
// Standalone Functions
//==================================================================================================

// Tests whether we can get file status information.
void test_stat(void)
{
    fprintf(stderr, "testing stat() ... ");

    const char *filename = "README.md";

    struct stat st = {0};

    // Get file status information and assert result.
    assert(stat(filename, &st) == 0);
    assert(st.st_mode & S_IFREG);
    assert(st.st_size > 0);
    assert(st.st_nlink > 0);
    assert(st.st_blocks > 0);
    assert(st.st_atim.tv_sec > 0);
    assert(st.st_mtim.tv_sec > 0);
    assert(st.st_ctim.tv_sec > 0);
    assert(st.st_atim.tv_nsec >= 0);
    assert(st.st_mtim.tv_nsec >= 0);
    assert(st.st_ctim.tv_nsec >= 0);
    assert(st.st_blocks > 0);
    assert(st.st_dev != 0);
    assert(st.st_ino != 0);
    assert(st.st_nlink > 0);

    fprintf(stderr, "passed\n");
}

File: src/tests/file-rust/src/unsafe/fs/stat.rs
// Copyright(c) The Maintainers of Nanvix.
// Licensed under the MIT License.

//==================================================================================================
// Imports
//==================================================================================================

use ::syscall::{
    sys,
    sys::stat::file_mode,
};

//==================================================================================================
// Standalone Functions
//==================================================================================================

/// Tests wether we can get the status of a file using `stat()`.
pub fn test() {
    let filename: &str = "README.md";

    // Get file status and assert result.
    let mut st: sys::stat::stat = sys::stat::stat::default();
    match sys::stat::stat(filename, &mut st) {
        Ok(()) => {
            // Check if the file is a regular file.
            if !file_mode::S_ISREG(st.st_mode) {
                panic!("file is not a regular file");
            }
        },
        Err(error) => {
            panic!("{error:?}");
        },
    }
}
